File: src/programmers/helpers/UnionPredicator.ts
import { Metadata } from "../../schemas/metadata/Metadata";
import { MetadataObjectType } from "../../schemas/metadata/MetadataObjectType";
import { MetadataProperty } from "../../schemas/metadata/MetadataProperty";

import { ArrayUtil } from "../../utils/ArrayUtil";
import { MapUtil } from "../../utils/MapUtil";

export namespace UnionPredicator {
  export interface ISpecialized {
    index: number;
    object: MetadataObjectType;
    property: MetadataProperty;
    neighbor: boolean;
  }

  export const object = (
    objects: MetadataObjectType[],
  ): Array<ISpecialized> => {
    // PROPERTY MATRIX
    const matrix: Map<string, Array<MetadataProperty | null>> = new Map();
    for (const obj of objects)
      for (const prop of obj.properties) {
        const key: string | null = prop.key.getSoleLiteral();
        if (key !== null)
          MapUtil.take(matrix, key, () =>
            ArrayUtil.repeat(objects.length, () => null),
          );
      }
    objects.forEach((obj, i) => {
      for (const prop of obj.properties) {
        const key: string | null = prop.key.getSoleLiteral();
        if (key !== null) matrix.get(key)![i] = prop;
      }
    });

    // EXPLORE SPECIALIZERS
    const output: ISpecialized[] = [];
    objects.forEach((obj, i) => {
      const children: ISpecializedProperty[] = [];
      obj.properties.forEach((prop) => {
        // MUST BE REQUIRED
        if (prop.value.isRequired() === false) return;
        const key: string | null = prop.key.getSoleLiteral();
        if (key === null) return;

        // FIND NEIGHBORHOOD PROPERTIES
        const neighbors: MetadataProperty[] = matrix
          .get(key)!
          .filter((oppo, k) => i !== k && oppo !== null) as MetadataProperty[];

        // NO NEIGHBORHOOD
        const unique: boolean =
          neighbors.length === 0 ||
          neighbors.every((n) => !Metadata.intersects(prop.value, n.value));
        if (unique === true)
          children.push({
            property: prop,
            neighbor: neighbors.length !== 0,
          });
      });
      if (children.length === 0) return;

      const top: ISpecializedProperty =
        children.find((child) => child.property.value.isConstant()) ||
        children[0]!;
      output.push({
        index: i,
        object: obj,
        ...top,
      });
    });
    return output;
  };
}

interface ISpecializedProperty {
  property: MetadataProperty;
  neighbor: boolean;
}

File: packages/typia/native/core/programmers/helpers/UnionPredicator.go
package helpers

import nativemetadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"

type unionPredicatorNamespace struct{}

var UnionPredicator = unionPredicatorNamespace{}

type UnionPredicator_ISpecialized struct {
  Index    int
  Object   *nativemetadata.MetadataObjectType
  Property *nativemetadata.MetadataProperty
  Neighbor bool
}

type unionPredicator_ISpecializedProperty struct {
  Property *nativemetadata.MetadataProperty
  Neighbor bool
}

type unionPredicator_propertyEntry struct {
  Property *nativemetadata.MetadataProperty
  Key      string
}

func (unionPredicatorNamespace) Object(objects []*nativemetadata.MetadataObjectType) []UnionPredicator_ISpecialized {
  matrix := map[string][]*nativemetadata.MetadataProperty{}
  entries := make([][]unionPredicator_propertyEntry, len(objects))
  for i, obj := range objects {
    entries[i] = make([]unionPredicator_propertyEntry, 0, len(obj.Properties))
    for _, prop := range obj.Properties {
      key := prop.Key.GetSoleLiteral()
      if key == nil {
        continue
      }
      text := *key
      if _, ok := matrix[text]; ok == false {
        matrix[text] = make([]*nativemetadata.MetadataProperty, len(objects))
      }
      matrix[text][i] = prop
      entries[i] = append(entries[i], unionPredicator_propertyEntry{
        Property: prop,
        Key:      text,
      })
    }
  }

  output := []UnionPredicator_ISpecialized{}
  for i, entry := range entries {
    children := []unionPredicator_ISpecializedProperty{}
    for _, item := range entry {
      prop := item.Property
      if prop.Value.IsRequired() == false {
        continue
      }
      neighbor := false
      unique := true
      for k, oppo := range matrix[item.Key] {
        if i != k && oppo != nil {
          neighbor = true
          if nativemetadata.MetadataSchema_intersects(prop.Value, oppo.Value) {
            unique = false
            break
          }
        }
      }
      if unique {
        children = append(children, unionPredicator_ISpecializedProperty{
          Property: prop,
          Neighbor: neighbor,
        })
      }
    }
    if len(children) == 0 {
      continue
    }
    top := children[0]
    for _, child := range children {
      if child.Property.Value.IsConstant() {
        top = child
        break
      }
    }
    output = append(output, UnionPredicator_ISpecialized{
      Index:    i,
      Object:   objects[i],
      Property: top.Property,
      Neighbor: top.Neighbor,
    })
  }
  return output
}
